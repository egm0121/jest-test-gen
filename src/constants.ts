export const FILE_SUFFIX = process.env.JEST_TEST_GEN_FILESUFFIX;
export const MOCK_MODULES_BLACKLIST = [
  'react',
  'prop-types',
  'styled-components',
];
export const IMPORT_MODULES_BLACKLIST = ['prop-types', 'styled-components'];
