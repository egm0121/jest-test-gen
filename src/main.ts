import { readFileSync, writeFileSync } from 'fs';
import * as path from 'path';
import * as ts from 'typescript';
import { parseSourceFile } from './parse-source-file';
import { generateUnitTest } from './generate-unit-test';

type TRunOptions = {
  returnOutput?: boolean
}

export function run(params: string[], opts?: TRunOptions) {
  if (!params.length) {
    // tslint:disable-next-line:no-console
    console.error('missing path argument');
    process.exit(1);
  }

  if (params.length > 1 && params[0].indexOf('--require') === 0) {
    require(params[1]);
    params = params.slice(2);
  }
  const inputPath = params[0];
  const inputFilenameNoExt = path.basename(inputPath, path.extname(inputPath));
  const specFileName = path.join(path.dirname(inputPath),`${inputFilenameNoExt}.generated.test.js`);

  const sourceCode = readFileSync(inputPath).toString();

  const sourceFile = ts.createSourceFile(
    inputPath,
    sourceCode,
    ts.ScriptTarget.Latest,
    true, /*setParentNodes */
  );
  const input = parseSourceFile(sourceFile);
  const output = generateUnitTest(inputPath, sourceCode, input);
  if(opts?.returnOutput){
    return output;
  }
  return writeFileSync(specFileName, output);
}