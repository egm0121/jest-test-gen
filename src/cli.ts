#!/usr/bin/env node
import * as minimist from 'minimist';
import { run } from './main';
import checkPackageUpdates from './updateVersionBanner';

checkPackageUpdates();
run(minimist(process.argv.slice(2)));
