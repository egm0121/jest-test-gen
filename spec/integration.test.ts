import { run } from '../src/main';

describe('integration', () => {
  describe(`single class default export by identifier`, () => {
    createSnapshotTest({
      file: 'spec/fixtures/classes/defaultExportIdentifier.js',
    });
  });
  describe(`single class default export by identifier with static methods`, () => {
    createSnapshotTest({
      file: 'spec/fixtures/classes/defaultExportStaticMethod.js',
    });
  });
  describe(`single class default export of assigned class`, () => {
    createSnapshotTest({
      file: 'spec/fixtures/classes/defaultExportAssignment.js',
    });
  });
  describe(`single class named export of assigned class definition`, () => {
    createSnapshotTest({
      file: 'spec/fixtures/classes/namedExportClassDefinition.js',
    });
  });
  describe(`single class named export of class identifier`, () => {
    createSnapshotTest({
      file: 'spec/fixtures/classes/namedExportIdentifier.js',
    });
  });
  describe(`single class named export of variable declaration with class assignement`, () => {
    createSnapshotTest({
      file: 'spec/fixtures/classes/namedExportVariableDeclaration.js',
    });
  });
  describe(`function default export assignement`, () => {
    createSnapshotTest({
      file: 'spec/fixtures/functions/defaultExportAssignement.js',
    });
  });
  describe(`function default export identifier`, () => {
    createSnapshotTest({
      file: 'spec/fixtures/functions/defaultExportIdentifier.js',
    });
  });
  describe(`function named export definition`, () => {
    createSnapshotTest({
      file: 'spec/fixtures/functions/namedExportDefinition.js',
    });
  });
  describe(`function named export variable statement`, () => {
    createSnapshotTest({
      file: 'spec/fixtures/functions/namedExportVariableStatement.js',
    });
  });
  describe(`pojo default export identifier`, () => {
    createSnapshotTest({
      file: 'spec/fixtures/pojos/defaultExportIdentifier.js',
    });
  });
  describe(`pojo named export assignement`, () => {
    createSnapshotTest({
      file: 'spec/fixtures/pojos/namedExportAssignement.js',
    });
  });
  describe(`pojo named export identifier`, () => {
    createSnapshotTest({
      file: 'spec/fixtures/pojos/namedExportIdentifier.js',
    });
  });
  describe(`component class based`, () => {
    createSnapshotTest({
      file: 'spec/fixtures/components/classComponent.js',
    });
  });
  describe(`component class based - default export`, () => {
    createSnapshotTest({
      file: 'spec/fixtures/components/defaultClassComponent.js',
    });
  });
  describe(`component functional - default export`, () => {
    createSnapshotTest({
      file: 'spec/fixtures/components/functionalDefaultExportComponent.js',
    });
  });
  describe(`component functional - anonymous default export`, () => {
    createSnapshotTest({
      file: 'spec/fixtures/components/unnamedDefaultComponent.js',
    });
  });
  describe(`component functional - named export`, () => {
    createSnapshotTest({
      file: 'spec/fixtures/components/functionalNamedComponent.js',
    });
  });
  describe(`components multiple - named export`, () => {
    createSnapshotTest({
      file: 'spec/fixtures/components/multipleComponent.js',
    });
  });
  describe(`typescript support`, () => {
    describe('component class based - generic arg parsing', () => {
      createSnapshotTest({
        file: 'spec/fixtures/components/tsFunctionalTyped.tsx'
      })
    });
    describe('component functional - typed prop argument', () => {
      createSnapshotTest({
        file: 'spec/fixtures/components/tsFunctionalTyped.tsx'
      })
    });
    describe('component functional - typed prop argument - array of type syntax', () => {
      createSnapshotTest({
        file: 'spec/fixtures/components/tsFunctionalTypedArray.tsx'
      })
    });
    describe('component functional - typed prop argument - function declaration parsing', () => {
      createSnapshotTest({
        file: 'spec/fixtures/components/tsFunctionalTypedDeclaration.tsx'
      })
    });
    describe('component functional - FunctionComponent<propType> - generic type arg parsing', () => {
      createSnapshotTest({
        file: 'spec/fixtures/components/tsFunctionalGeneric.tsx'
      })
    });
    describe('component functional - React.FC<propType> - generic type arg parsing', () => {
      createSnapshotTest({
        file: 'spec/fixtures/components/tsFunctionalGeneric_React.FC.tsx'
      })
    });
  });
});

function createSnapshotTest(input: {
  file: string,
}) {
  it('should match snapshot', () => {
    expect(run({ _: [input.file] }, { returnOutput: true })).toMatchSnapshot();
  });
}
