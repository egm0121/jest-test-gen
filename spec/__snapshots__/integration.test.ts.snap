// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration component class based - default export should match snapshot 1`] = `
"import renderer from 'react-test-renderer';
import React from 'react';
import MyTestComponent from './defaultClassComponent';

const renderTree = tree => renderer.create(tree);
describe('<MyTestComponent>', () => {
  it('should render component', () => {
    expect(renderTree(<MyTestComponent  
      magicNumber={/*  PropTypes.number.isRequired */} 
    />).toJSON()).toMatchSnapshot();
  });
  it('should render component with props', () => {
    expect(renderTree(<MyTestComponent  
      magicNumber={/*  PropTypes.number.isRequired */}  
      optional={/*  PropTypes.string */}  
      style={/*  React.CSSProperties */} 
    />).toJSON()).toMatchSnapshot();
  });
});"
`;

exports[`integration component class based should match snapshot 1`] = `
"import renderer from 'react-test-renderer';
import React from 'react';
import { connect } from 'react-redux';
import { CallTimer } from './classComponent';

jest.mock('react-redux');

const renderTree = tree => renderer.create(tree);
describe('<CallTimer>', () => {
  it('should render component', () => {
    expect(renderTree(<CallTimer  
      magicNumber={/*  PropTypes.number.isRequired */} 
    />).toJSON()).toMatchSnapshot();
  });
  it('should render component with props', () => {
    expect(renderTree(<CallTimer  
      magicNumber={/*  PropTypes.number.isRequired */}  
      optional={/*  PropTypes.string */}  
      style={/*  React.CSSProperties */} 
    />).toJSON()).toMatchSnapshot();
  });
});"
`;

exports[`integration component functional - anonymous default export should match snapshot 1`] = `
"import renderer from 'react-test-renderer';
import React from 'react';
import UnnamedDefaultComponent from './unnamedDefaultComponent';

const renderTree = tree => renderer.create(tree);
describe('<UnnamedDefaultComponent>', () => {
  it('should render component', () => {
    expect(renderTree(<UnnamedDefaultComponent 
    />).toJSON()).toMatchSnapshot();
  });
  
});"
`;

exports[`integration component functional - default export should match snapshot 1`] = `
"import renderer from 'react-test-renderer';
import React from 'react';
import MyDefaultComponent from './functionalDefaultExportComponent';

const renderTree = tree => renderer.create(tree);
describe('<MyDefaultComponent>', () => {
  it('should render component', () => {
    expect(renderTree(<MyDefaultComponent 
    />).toJSON()).toMatchSnapshot();
  });
  it('should render component with props', () => {
    expect(renderTree(<MyDefaultComponent  
      message={/*  PropTypes.string */} 
    />).toJSON()).toMatchSnapshot();
  });
});"
`;

exports[`integration component functional - named export should match snapshot 1`] = `
"import renderer from 'react-test-renderer';
import React from 'react';
import { MyFComponent } from './functionalNamedComponent';

const renderTree = tree => renderer.create(tree);
describe('<MyFComponent>', () => {
  it('should render component', () => {
    expect(renderTree(<MyFComponent 
    />).toJSON()).toMatchSnapshot();
  });
  it('should render component with props', () => {
    expect(renderTree(<MyFComponent  
      message={/*  PropTypes.string */} 
    />).toJSON()).toMatchSnapshot();
  });
});"
`;

exports[`integration components multiple - named export should match snapshot 1`] = `
"import renderer from 'react-test-renderer';
import React from 'react';
import MyTestComponent, { SmallComponent } from './multipleComponent';

const renderTree = tree => renderer.create(tree);
describe('<SmallComponent>', () => {
  it('should render component', () => {
    expect(renderTree(<SmallComponent 
    />).toJSON()).toMatchSnapshot();
  });
  
});
describe('<MyTestComponent>', () => {
  it('should render component', () => {
    expect(renderTree(<MyTestComponent  
      magicNumber={/*  PropTypes.number.isRequired */} 
    />).toJSON()).toMatchSnapshot();
  });
  it('should render component with props', () => {
    expect(renderTree(<MyTestComponent  
      magicNumber={/*  PropTypes.number.isRequired */}  
      optional={/*  PropTypes.string */}  
      style={/*  React.CSSProperties */} 
    />).toJSON()).toMatchSnapshot();
  });
});"
`;

exports[`integration function default export assignement should match snapshot 1`] = `
"import moduleOne from 'moduleOne';
import moduleTwo from 'moduleTwo';
import myDefaultExport, { myNamedExport } from 'moduleThree/deep/file';
import helpers from './local/helpers';
import defaultExportAssignement from './defaultExportAssignement';

jest.mock('moduleOne');
jest.mock('moduleTwo');
jest.mock('moduleThree/deep/file');
jest.mock('./local/helpers');

describe('defaultExportAssignement', () => {
  it('should expose a function', () => {
		expect(defaultExportAssignement).toBeDefined();
	});
  
  it('defaultExportAssignement should return expected output', () => {
    // const retValue = defaultExportAssignement(myArgOne,myArgTwo);
    expect(false).toBeTruthy();
  });
});"
`;

exports[`integration function default export identifier should match snapshot 1`] = `
"import moduleOne from 'moduleOne';
import moduleTwo from 'moduleTwo';
import myDefaultExport, { myNamedExport } from 'moduleThree/deep/file';
import helpers from './local/helpers';
import helloDefFunc from './defaultExportIdentifier';

jest.mock('moduleOne');
jest.mock('moduleTwo');
jest.mock('moduleThree/deep/file');
jest.mock('./local/helpers');

describe('helloDefFunc', () => {
  it('should expose a function', () => {
		expect(helloDefFunc).toBeDefined();
	});
  
  it('helloDefFunc should return expected output', async () => {
    // const retValue = await helloDefFunc();
    expect(false).toBeTruthy();
  });
});"
`;

exports[`integration function named export definition should match snapshot 1`] = `
"import moduleOne from 'moduleOne';
import moduleTwo from 'moduleTwo';
import myDefaultExport, { myNamedExport } from 'moduleThree/deep/file';
import helpers from './local/helpers';
import { helloNamedFunction } from './namedExportDefinition';

jest.mock('moduleOne');
jest.mock('moduleTwo');
jest.mock('moduleThree/deep/file');
jest.mock('./local/helpers');

describe('helloNamedFunction', () => {
  it('should expose a function', () => {
		expect(helloNamedFunction).toBeDefined();
	});
  
  it('helloNamedFunction should return expected output', () => {
    // const retValue = helloNamedFunction(myArgOne,myArgTwo);
    expect(false).toBeTruthy();
  });
});"
`;

exports[`integration function named export variable statement should match snapshot 1`] = `
"import moduleOne from 'moduleOne';
import moduleTwo from 'moduleTwo';
import myDefaultExport, { myNamedExport } from 'moduleThree/deep/file';
import helpers from './local/helpers';
import { helloAssignedArrowFunction, myAsyncAssignedFunc } from './namedExportVariableStatement';

jest.mock('moduleOne');
jest.mock('moduleTwo');
jest.mock('moduleThree/deep/file');
jest.mock('./local/helpers');

describe('helloAssignedArrowFunction', () => {
  it('should expose a function', () => {
		expect(helloAssignedArrowFunction).toBeDefined();
	});
  
  it('helloAssignedArrowFunction should return expected output', () => {
    // const retValue = helloAssignedArrowFunction(myArgOne,myArgTwo);
    expect(false).toBeTruthy();
  });
});
describe('myAsyncAssignedFunc', () => {
  it('should expose a function', () => {
		expect(myAsyncAssignedFunc).toBeDefined();
	});
  
  it('myAsyncAssignedFunc should return expected output', async () => {
    // const retValue = await myAsyncAssignedFunc(arg1);
    expect(false).toBeTruthy();
  });
});"
`;

exports[`integration pojo default export identifier should match snapshot 1`] = `
"import moduleOne from 'moduleOne';
import moduleTwo from 'moduleTwo';
import myDefaultExport, { myNamedExport } from 'moduleThree/deep/file';
import helpers from './local/helpers';
import myPojo from './defaultExportIdentifier';

jest.mock('moduleOne');
jest.mock('moduleTwo');
jest.mock('moduleThree/deep/file');
jest.mock('./local/helpers');

describe('myPojo', () => {
  
  it('should expose a method methodOne()', () => {
    //const retValue = myPojo.methodOne(param1);
    expect(false).toBeTruthy();
  });
  
  it('should expose a method methodTwo()', async () => {
    //const retValue = await myPojo.methodTwo(a,b);
    expect(false).toBeTruthy();
  });
  
});"
`;

exports[`integration pojo named export assignement should match snapshot 1`] = `
"import moduleOne from 'moduleOne';
import moduleTwo from 'moduleTwo';
import myDefaultExport, { myNamedExport } from 'moduleThree/deep/file';
import helpers from './local/helpers';
import { myPojo } from './namedExportAssignement';

jest.mock('moduleOne');
jest.mock('moduleTwo');
jest.mock('moduleThree/deep/file');
jest.mock('./local/helpers');

describe('myPojo', () => {
  
  it('should expose a method methodOne()', () => {
    //const retValue = myPojo.methodOne(param1);
    expect(false).toBeTruthy();
  });
  
  it('should expose a method methodTwo()', async () => {
    //const retValue = await myPojo.methodTwo(a,b);
    expect(false).toBeTruthy();
  });
  
});"
`;

exports[`integration pojo named export identifier should match snapshot 1`] = `
"import moduleOne from 'moduleOne';
import moduleTwo from 'moduleTwo';
import myDefaultExport, { myNamedExport } from 'moduleThree/deep/file';
import helpers from './local/helpers';
import { myPojo } from './namedExportIdentifier';

jest.mock('moduleOne');
jest.mock('moduleTwo');
jest.mock('moduleThree/deep/file');
jest.mock('./local/helpers');

describe('myPojo', () => {
  
  it('should expose a method methodOne()', () => {
    //const retValue = myPojo.methodOne(param1);
    expect(false).toBeTruthy();
  });
  
  it('should expose a method methodTwo()', async () => {
    //const retValue = await myPojo.methodTwo(a,b);
    expect(false).toBeTruthy();
  });
  
});"
`;

exports[`integration single class default export by identifier should match snapshot 1`] = `
"import moduleOne from 'moduleOne';
import moduleTwo from 'moduleTwo';
import myDefaultExport, { myNamedExport } from 'moduleThree/deep/file';
import helpers from './local/helpers';
import MyDemoClass from './defaultExportIdentifier';

jest.mock('moduleOne');
jest.mock('moduleTwo');
jest.mock('moduleThree/deep/file');
jest.mock('./local/helpers');

describe('MyDemoClass', () => {
  let instance;

  beforeEach(() => {
    instance = new MyDemoClass();
  });

  it('instance should be an instanceof MyDemoClass', () => {
    expect(instance instanceof MyDemoClass).toBeTruthy();
  });

  it('should have a method hiThere()', () => {
    // instance.hiThere();
    expect(false).toBeTruthy();
  });

  it('should have a method myDemoMethod()', () => {
    // instance.myDemoMethod();
    expect(false).toBeTruthy();
  });
});"
`;

exports[`integration single class default export by identifier with static methods should match snapshot 1`] = `
"import moduleOne from 'moduleOne';
import moduleTwo from 'moduleTwo';
import myDefaultExport, { myNamedExport } from 'moduleThree/deep/file';
import helpers from './local/helpers';
import DefaultExportStaticMethod from './defaultExportStaticMethod';

jest.mock('moduleOne');
jest.mock('moduleTwo');
jest.mock('moduleThree/deep/file');
jest.mock('./local/helpers');

describe('DefaultExportStaticMethod', () => {
  let instance;

  beforeEach(() => {
    instance = new DefaultExportStaticMethod();
  });

  it('instance should be an instanceof DefaultExportStaticMethod', () => {
    expect(instance instanceof DefaultExportStaticMethod).toBeTruthy();
  });

  it('should have a static method makeInstance()', () => {
    // DefaultExportStaticMethod.makeInstance(config);
    expect(false).toBeTruthy();
  });

  it('should have a method myDemoMethod()', () => {
    // instance.myDemoMethod();
    expect(false).toBeTruthy();
  });
});"
`;

exports[`integration single class default export of assigned class should match snapshot 1`] = `
"import moduleOne from 'moduleOne';
import moduleTwo from 'moduleTwo';
import myDefaultExport, { myNamedExport } from 'moduleThree/deep/file';
import helpers from './local/helpers';
import DefaultExportAssignment from './defaultExportAssignment';

jest.mock('moduleOne');
jest.mock('moduleTwo');
jest.mock('moduleThree/deep/file');
jest.mock('./local/helpers');

describe('DefaultExportAssignment', () => {
  let instance;

  beforeEach(() => {
    instance = new DefaultExportAssignment();
  });

  it('instance should be an instanceof DefaultExportAssignment', () => {
    expect(instance instanceof DefaultExportAssignment).toBeTruthy();
  });

  it('should have a method hiThere()', () => {
    // instance.hiThere();
    expect(false).toBeTruthy();
  });

  it('should have a method myDemoMethod()', () => {
    // instance.myDemoMethod();
    expect(false).toBeTruthy();
  });
});"
`;

exports[`integration single class named export of assigned class definition should match snapshot 1`] = `
"import moduleOne from 'moduleOne';
import moduleTwo from 'moduleTwo';
import myDefaultExport, { myNamedExport } from 'moduleThree/deep/file';
import helpers from './local/helpers';
import { MyDemoClass, myTestHelper } from './namedExportClassDefinition';

jest.mock('moduleOne');
jest.mock('moduleTwo');
jest.mock('moduleThree/deep/file');
jest.mock('./local/helpers');

describe('MyDemoClass', () => {
  let instance;

  beforeEach(() => {
    instance = new MyDemoClass();
  });

  it('instance should be an instanceof MyDemoClass', () => {
    expect(instance instanceof MyDemoClass).toBeTruthy();
  });

  it('should have a method hiThere()', () => {
    // instance.hiThere(name,surname);
    expect(false).toBeTruthy();
  });

  it('should have a method myDemoMethodOne()', () => {
    // instance.myDemoMethodOne(demoParam);
    expect(false).toBeTruthy();
  });

  it('should have a method myDemoMethodTwo()', () => {
    // instance.myDemoMethodTwo(demoParam);
    expect(false).toBeTruthy();
  });

  it('should have a method myDemoMethodThree()', () => {
    // instance.myDemoMethodThree(demoParam);
    expect(false).toBeTruthy();
  });
});

describe('myTestHelper', () => {
  it('should expose a function', () => {
		expect(myTestHelper).toBeDefined();
	});
  
  it('myTestHelper should return expected output', () => {
    // const retValue = myTestHelper(arg1,arg2);
    expect(false).toBeTruthy();
  });
});"
`;

exports[`integration single class named export of class identifier should match snapshot 1`] = `
"import moduleOne from 'moduleOne';
import moduleTwo from 'moduleTwo';
import myDefaultExport, { myNamedExport } from 'moduleThree/deep/file';
import helpers from './local/helpers';
import { MyDemoClass } from './namedExportIdentifier';

jest.mock('moduleOne');
jest.mock('moduleTwo');
jest.mock('moduleThree/deep/file');
jest.mock('./local/helpers');

describe('MyDemoClass', () => {
  let instance;

  beforeEach(() => {
    instance = new MyDemoClass();
  });

  it('instance should be an instanceof MyDemoClass', () => {
    expect(instance instanceof MyDemoClass).toBeTruthy();
  });

  it('should have a method hiThere()', () => {
    // instance.hiThere();
    expect(false).toBeTruthy();
  });

  it('should have a method myDemoMethod()', () => {
    // instance.myDemoMethod();
    expect(false).toBeTruthy();
  });
});"
`;

exports[`integration single class named export of variable declaration with class assignement should match snapshot 1`] = `
"import moduleOne from 'moduleOne';
import moduleTwo from 'moduleTwo';
import myDefaultExport, { myNamedExport } from 'moduleThree/deep/file';
import helpers from './local/helpers';
import { MyVarClass } from './namedExportVariableDeclaration';

jest.mock('moduleOne');
jest.mock('moduleTwo');
jest.mock('moduleThree/deep/file');
jest.mock('./local/helpers');

describe('MyVarClass', () => {
  let instance;

  beforeEach(() => {
    instance = new MyVarClass();
  });

  it('instance should be an instanceof MyVarClass', () => {
    expect(instance instanceof MyVarClass).toBeTruthy();
  });

  it('should have a method hiThere()', () => {
    // instance.hiThere();
    expect(false).toBeTruthy();
  });

  it('should have a method myDemoMethod()', () => {
    // instance.myDemoMethod(myMethodArg1);
    expect(false).toBeTruthy();
  });
});"
`;
